AutoCollection/HeartBeat.ts

```diff
@@ -0,0 +1,138 @@
+
+    private static AIMS_URI = "http://169.254.169.254/metadata/instance/compute";
+    private static AIMS_API_VERSION = "api-version=2017-12-01";
+    private static AIMS_FORMAT = "format=json";
+
+    private _collectionInterval: number = 900000;
+    private _client: TelemetryClient;
+    private _handle: NodeJS.Timer;
+    private _isEnabled: boolean;
+    private _isInitialized: boolean;
+    private _isVM: boolean = false;
+    private _vmData = <{[key: string]: string}>{};
+
+    constructor(client: TelemetryClient) {
+        if (!HeartBeat.INSTANCE) {
+            HeartBeat.INSTANCE = this;
+        }
+
+        this._isInitialized = false;
+        this._client = client;
+    }
+
+    public enable(isEnabled: boolean, config?: Config) {
+        this._isEnabled = isEnabled;
+        if (this._isEnabled && !this._isInitialized) {
+            this._isInitialized = true;
+        }
+
+        if (isEnabled) {
+            if (!this._handle) {
+                this._handle = setInterval(() => this.trackHeartBeat(config, () => {}), this._collectionInterval);
+                this._handle.unref(); // Allow the app to terminate even while this loop is going on
+            }
+        } else {
+            if (this._handle) {
+                clearInterval(this._handle);
+                this._handle = undefined;
+            }
+        }
+    }
+
+    public isInitialized() {
+        return this._isInitialized;
+    }
+
+    public static isEnabled() {
+        return HeartBeat.INSTANCE && HeartBeat.INSTANCE._isEnabled;
+    }
+
+    public trackHeartBeat(config: Config, callback: () => void) {
+        let waiting: boolean = false;
+        let properties: {[key: string]: string} = {};
+        const sdkVersion = Context.sdkVersion; // "node" or "node-nativeperf"
+        properties["sdk"] = sdkVersion;
+        properties["osType"] = os.type();
+        if (process.env.WEBSITE_SITE_NAME) { // Web apps
+            properties["appSrv_SiteName"] = process.env.WEBSITE_SITE_NAME || "";
+            properties["appSrv_wsStamp"] = process.env.WEBSITE_HOME_STAMPNAME || "";
+            properties["appSrv_wsHost"] = process.env.WEBSITE_HOSTNAME || "";
+        } else if (process.env.FUNCTIONS_WORKER_RUNTIME) { // Function apps
+            properties["azfunction_appId"] = process.env.WEBSITE_HOSTNAME;
+        } else if (config) { // VM
+            waiting = true;
+            this._getAzureComputeMetadata(config, () => {
+                if (this._isVM && Object.keys(this._vmData).length > 0) {
+                    properties["azInst_vmId"] = this._vmData["vmId"] || "";
+                    properties["azInst_subscriptionId"] = this._vmData["subscriptionId"] || "";
+                    properties["azInst_osType"] = this._vmData["osType"] || "";
+                }
+                this._client.trackMetric({name: Constants.HeartBeatMetricName, value: 0, properties: properties});
+                callback();
```

Comment:
Nit: Why not just have `callback()` called at the end instead instead of in these if blocks?

Reply:
Because checking for environment variables no need to call callback.